feat(person-on-events): add option to delay all events

hazzadous opened the pull request:
This change implements the option outlined in
https://github.com/PostHog/product-internal/pull/405

Here I do not try to do any large structural changes to the code, I'll
leave that for later although it does mean the code has a few loose
couplings between pipeline steps that probably should be strongly
coupled. I've tried to comment these to try to make it clear about the
couplings.

I've also added a workflow to run the functional tests against both
configurations, which we can remove once we're happy with the new
implementation.

Things of note:

 1. We can't enable this for all users yet, not without the live events
    view and not without verifying that the buffer size is sufficiently
    large. We can however enable this for the test team and verify that
    it functions as expected.
 2. I have not handled the case mentioned in the above PR regarding
    guarding against processing the delayed events before all events in
    the delay window have been processed.

TODO: decide what should happen with `onEvent`/`exportEvents`. i.e. when should these be sent. Some people require that these be sent immediately.

## Problem

<!-- Who are we building for, what are their needs, why is this important? -->

## Changes

<!-- If there are frontend changes, please include screenshots. -->
<!-- If a reference design was involved, include a link to the relevant Figma frame! -->

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

Commit:
feat(person-on-events): add option to delay all events

This change implements the option outlined in
https://github.com/PostHog/product-internal/pull/405

Here I do not try to do any large structural changes to the code, I'll
leave that for later although it does mean the code has a few loose
couplings between pipeline steps that probably should be strongly
coupled. I've tried to comment these to try to make it clear about the
couplings.

I've also added a workflow to run the functional tests against both
configurations, which we can remove once we're happy with the new
implementation.

Things of note:

 1. We can't enable this for all users yet, not without the live events
    view and not without verifying that the buffer size is sufficiently
    large. We can however enable this for the test team and verify that
    it functions as expected.
 2. I have not handled the case mentioned in the above PR regarding
    guarding against processing the delayed events before all events in
    the delay window have been processed.

wip

test(person-on-events): add currently failing test for person on events

This test doesn't work with the previous behaviour of the
person-on-events implementation, but should pass with the new delay all
events behaviour.

Commit:
add test for KafkaJSError behaviour

Commit:
add comment re delay

Commit:
add test for create_alias

Commit:
chore: increase exports timeout

It seems to fail in CI, but only for the delayed events enabled tests.
I'm not sure why, but I'm guessing it's because the events are further
delayed by the new implementation.

Commit:
chore: fix test

Commit:
add test for ordering of person properties

Commit:
use ubuntu-latest-8-cores runner

Commit:
add tests for plugin processEvent

Commit:
chore: ensure plugin processEvent isn't run multiple times

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -90,6 +90,12 @@ export class PersonState {
     }
 
+    async updateExceptProperties(): Promise<LazyPersonContainer> {
+        await this.handleIdentifyOrAlias()
```
Note that this merges properties mentioned in the RFC PR too, Ideally we'd only link the distinct_ids and merged the properties later, however that's the "fun" code we have, so maybe we can ignore this for now 🤷‍♀️

hazzadous replied:
Ok, I'll add a test for this case also.

hazzadous replied:
@tiina303 I'd expanded on the person properties test, it doesn't change the code however. Do you have a test scenario that would fail as is?

tiina303 replied:
Example scenario that would be different if we do properties merge before vs after:
```
identify(alice, {k: v1})
identify(bob, {k: v2})  -- or just identify(bob)
-- wait until processed (or just set up the persons that way before the test)
capture(bob, properties={$set: {k: v3}})  
create_alias(alice, bob)
-- these two would be in the buffer window, so 
1. if we process property merges early, we'll end up with the merge first which takes k=v1 from Alice (we merge into the first person), then process the property from identify, then we end up with overriding the value from the capture, i.e. getting k = v3
2. if we do property merges later, then we'll merge Alice with k=v1 and bob with k=v3 into Alice (keeping the value on conflict from the first person) we end up with k=v1
```

hazzadous replied:
Given that alice and bob are both "identified", I believe the alias will not work in this case. Should it work? I get this message:

```
🤔 refused to merge an already identified user via an $identify call
```

That's with the addition of this test: https://github.com/PostHog/posthog/pull/13505/commits/771d4b3f7d69780dfb01f9e0392d96627d0d8645

tiina303 replied:
aah well you can just do `custom events` for the first two too, the important thing was that they aren't in the buffer window

hazzadous replied:
Does this make sense now: https://github.com/PostHog/posthog/pull/13505/files#diff-e8c057ea332fa67a8f2bc6d811629e71d31b29630bf04fd906509be89f53e398R524

Commit:
expand on person properties ordering test

Commit:
wip

Commit:
wip

tiina303 commented on plugin-server/src/worker/ingestion/event-pipeline/runner.ts:
```diff
@@ -66,10 +66,12 @@ const STEPS_TO_EMIT_TO_DLQ_ON_FAILURE: Array<StepType> = [
 export class EventPipelineRunner {
     hub: Hub
     originalEvent: PipelineEvent | ProcessedPluginEvent
+    fullyProcessEvent: boolean
```
maybe call this `bufferAfterPersonCreation` (& flipped values)

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -90,6 +90,12 @@ export class PersonState {
         return this.personContainer
     }
 
+    async updateExceptProperties(): Promise<LazyPersonContainer> {
```
Can you add a comment here that we do process properties: if we are merging persons or if this creates the person (it could be very confusing to read/debug otherwise later)

tiina303 commented on plugin-server/src/worker/ingestion/person-state.ts:
```diff
@@ -90,6 +90,12 @@ export class PersonState {
 
+    async updateExceptProperties(): Promise<LazyPersonContainer> {
+        await this.handleIdentifyOrAlias()
+        await this.createPersonIfDistinctIdIsNew()
```
This also sets the properties for the person, which can cause things to be slightly odd too potentially

tiina303 replied:
Thinking of tests ... 
```
initial persons state: Alice k=v1 ; (can be separate test) Charlie: c = c1 ; Eve: c = e1 
-- these within buffer window:
capture(bob, properties={$set: {k: v3}})  -- creates bob
create_alias(alice, bob)
capture(Eve, properties={$set: {c: e2}})
create_alias(Charlie, Eve)
```
Here are the different options of processing setup: 
1. normal props late, both creation and merges early: k = v1 (because we first process Bob's creation, then merge) ; c = e2
1. normal props late, creation early, merges late (same as everything early / everything late as we can't have a test where we merge, then create as the person as merge would link the ids) k = v1 ; c = c1
1. normal props late, merges early, creation late:  k = v3 (because Bob gets created, but params processed after) ; c = e2

(normal props early, others late isn't an option we'd consider)

hazzadous replied:
I think I have a test that works for merges [here](https://github.com/PostHog/posthog/pull/13505/files#diff-e8c057ea332fa67a8f2bc6d811629e71d31b29630bf04fd906509be89f53e398R484)

I can't think of a test that fails for the `createPersonIfDistinctIdIsNew` adding properties out of order case, but I've updated to add a param to not update properties so finders crossed that works.

Commit:
add additional test

Commit:
change fullyProcessEvent to onlyUpdatePersonIdAssociations

Commit:
update test

tiina303 commented on plugin-server/src/worker/ingestion/event-pipeline/2-emitToBufferStep.ts:
```diff
@@ -24,6 +24,28 @@ export async function emitToBufferStep(
+        // TODO: remove this conditional once we're confident that the new
+        // ingestion pipeline is working well for all teams.
+        runner.onlyUpdatePersonIdAssociations = true
+        return runner.nextStep('pluginsProcessEventStep', event, personContainer)
```
It could be simpler if we did this instead:
2. plugins processing (so we always process plugins before buffer, could be useful for events that are dropped anyway to not go to the buffer & the person state here would be inconsistent for properties)
3. (potentially person ID handling) + emit to buffer
4. person handling (from buffer)

fuziontech replied:
This is a good idea but I think if I was using a plugin I'd prefer some lag + a better match on person than just running the plugin asap. This might be something where down the line we are reverse etl pushing to another service and to move a person from step 1 to 2 and having the right person would be really important there.

hazzadous replied:
I have [tests for two cases](https://github.com/PostHog/posthog/pull/13505/files#diff-4b66d551fbef355d8bf7d09eb82c707b0523fe08c2c8250fe41f08aa0b94df8dR119):

 1. plugins should be able to update attributes that person processing is dependent on
 2. if a plugin drops an event, it should not have any person side-effects

Happy to drop these requirements as desired.

fuziontech commented on plugin-server/functional_tests/plugins.test.ts:
```diff
@@ -90,16 +91,18 @@ test.concurrent(`plugin method tests: event captured, processed, ingested`, asyn
-        return events
+        expect(events[0].properties).toEqual(
+            expect.objectContaining({
+                processed: 'hell yes',
```
🤘

fuziontech reviewed (approved):
Amazing tests and comments as always. I'd follow up with Tiina's comments, but everything else looks good to me.

🚢 it

Commit:
add test to ensure person properties do not propagate backwards in time

Commit:
simplicfy person property tests

Commit:
weaken guarantee in test

Commit:
chore: make sure we don't update properties on the first parse

We should only be updating person_id and asociated distinct_ids on first
parse.

Commit:
add tests for dropping events

Commit:
increase export timeout

Commit:
increase historical exports timeout

Commit:
increase default waitForExpect interval to 1 second
